### PR TITLE
Be more strict about converting float to double

### PR DESCRIPTION
**Pull request by sw**
This enables `-Wdouble-promotion` and syncs the `Makefile` and `CMakeLists.txt` with regards to warnings.

Reasoning:
The llama.cpp codebase depends on the correct use of number types, whether those are `float`, `double` or some of the spezialized types such as q4_0. Converting from one type to another should be a concious decision and not happen by accident.

In order to avoid any type promotion warnings, I have updated the code base, sometimes by making an implicit cast explicit, but in some places I did change the actual semantics. I'm not confident at this point that all changes are good.

Consequences:
* Inference output changes, though I wouldn't say for the worse. Perplexity has an ETA of 20 hours on my machine, so I haven't run that yet.
* q4_0 quantization is identical.

Further steps if and when this PR is merged:
* Enable `-Werror`?

### CODE REVIEW

File: ggml.c [hunk 1]
```diff
@@ -1348,11 +1348,11 @@ inline static void ggml_vec_dot_f16(const int n, float * restrict s, ggml_fp16_t
 
     // leftovers
     for (int i = np; i < n; ++i) {
-        sumf += GGML_FP16_TO_FP32(x[i])*GGML_FP16_TO_FP32(y[i]);
+        sumf += (ggml_float)(GGML_FP16_TO_FP32(x[i])*GGML_FP16_TO_FP32(y[i]));
```
Comment (Green-Sky) [inline ggml.c:1351]: do you guys prefer `(int)(var1+var2)` or `int(var1+var2)` ?

File: ggml.c [hunk 2]
```diff
@@ -2091,11 +2091,11 @@ inline static void ggml_vec_sgn_f32  (const int n, float * y, const float * x) {
 inline static void ggml_vec_step_f32 (const int n, float * y, const float * x) { for (int i = 0; i < n; ++i) y[i] = (x[i] > 0.f) ? 1.f : 0.f; }
 inline static void ggml_vec_relu_f32 (const int n, float * y, const float * x) { for (int i = 0; i < n; ++i) y[i] = (x[i] > 0.f) ? x[i] : 0.f; }
 
-static const ggml_float GELU_COEF_A    = 0.044715;
-static const ggml_float SQRT_2_OVER_PI = 0.79788456080286535587989211986876;
+static const float GELU_COEF_A    = 0.044715;
+static const float SQRT_2_OVER_PI = 0.79788456080286535587989211986876;
```
Comment (Green-Sky) [inline ggml.c:2095]: should there be a `f` at the literal, to denote the type?

**Comment by Green-Sky** [comment]
i assume inference speed changes will be minimal, and only really a thing with simd disabled?

**Comment by anzz1** [comment]
I somewhat agree with:
- Typecastings **in most cases** should be explicit to denote conscious choices, however I understand that there can be a preference of not using them when you understand what you are doing (explicit *can be* easier for others to read and understand, but not always as in longer formulas it also can make it *harder* to read, and it *always* requires more typing)
- 1.0D and 1.0f is better than 1.0

However, I am not qualified to comment on the math itself. I can only say that changes like these require extra scrutiny, since having correct rounding is an integral part for the models to work properly. `round` != `roundf` , `sqrt` != `sqrtf`, etc. It is important to acknowledge the difference between number types, rounding and truncation including their effects on the results and speed of the calculations at the bare metal level.

---

One example:

https://github.com/ggerganov/llama.cpp/pull/458/commits/3afbe430935944fd936ca7b56b6e3219e43eebde#diff-6d9ce99fcb6f51ff76f59e479f6e6fc0bb62edef7442805d7a5bb15b23996b5dL482

To me it seems that calculations like this are very deliberate
```c
const float v0 = x[i*QK + l + 0]*id;
const float v1 = x[i*QK + l + 1]*id;
const uint8_t vi0 = ((int8_t) (round(v0))) + 8;
const uint8_t vi1 = ((int8_t) (round(v1))) + 8;
```

and to correctly explicitly cast it would be
```c
const uint8_t vi0 = (uint8_t)(((int8_t)(round((double)v0))) + 8);
```
this already shows why explicit typecastings can be more trouble than worth in many cases

however the way you changed it to
```c
const uint8_t vi0 = (int8_t)roundf(v0) + 8;
```
completely changes how the rounding and truncation works (+ the typecast still isn't correct) and require extra scrutiny and dialog between who originally wrote those calculations and anyone changing them.

Like I said, I am not qualified enough to comment on the math itself. I am just advising anyone merging any changes to the calculations until the changes being made and the effects they have are completely understood.

I am also not 100% certain whether this is a change which should go forward, but I'm not the one writing the formulas. As shown in the example above, explicit type-casting can also make longer formulas which include typecastings/truncations more obtuse by introducing long chains of extra types and parentheses.

This is just my few cents. In any case, this should be a dialog between the people writing the formulas. It should be their choice on how to proceed.

In any case, no decision should be based on such a superfluous thing as a compiler 'warning' you not to do something. You (should) understand better what you are doing than the compiler does. Therefore, for adding `-Werror` , my vote is a hard and resounding no. Unfortunately there isn't a perfect solution for this since not only the support for `#pragma warning( push )` and pops and suppresses vary between compilers but they also make the code unnecessarily spammier. As there are lot of valid use cases for code which generate warnings and some of the warnings are even downright stupid (like unused-result), the best way is simply being mindful about warnings in *your* code.

**Comment by anzz1** [comment]
> ... but not in this specific case. To prove it, I added a test that will loop through all possible floats (except NaNs and infinities) and verifies that the result is the same.

When you're right, you're right. In this case of just 256 possible values any rounding errors do not materialize. Yeah, I shouldn't comment on the math stuff. :smile: 

To be honest there is room for improvement, there are some questionable choices like `typedef ggml_float double;` which are a bit detrimental to readability but then again maybe there is a reason for it like a future expansion on mind or just something I don't get. 

Anyway good work and I hope this will get attention from the maths wizards.

**Review by ggerganov — APPROVED** [review]
This is a good change. Should have merged it before the refactoring.
If you resolve the conflicts, we can merge it

`-Werror` is a great idea, but too early to add it. Want to get to a more stable state first

File: examples/common.cpp [hunk 3]
```diff
@@ -206,13 +206,13 @@ void gpt_print_usage(int /*argc*/, char ** argv, const gpt_params & params) {
...
-    fprintf(stderr, "  --top_p N             top-p sampling (default: %.1f)\n", params.top_p);
+    fprintf(stderr, "  --top_p N             top-p sampling (default: %.1f)\n", (double)params.top_p);
     fprintf(stderr, "  --repeat_last_n N     last n tokens to consider for penalize (default: %d)\n", params.repeat_last_n);
-    fprintf(stderr, "  --repeat_penalty N    penalize repeat sequence of tokens (default: %.1f)\n", params.repeat_penalty);
+    fprintf(stderr, "  --repeat_penalty N    penalize repeat sequence of tokens (default: %.1f)\n", (double)params.repeat_penalty);
     fprintf(stderr, "  -c N, --ctx_size N    size of the prompt context (default: %d)\n", params.n_ctx);
     fprintf(stderr, "  --ignore-eos          ignore end of stream token and continue generating\n");
     fprintf(stderr, "  --memory_f32          use f32 instead of f16 for memory key+value\n");
-    fprintf(stderr, "  --temp N              temperature (default: %.1f)\n", params.temp);
+    fprintf(stderr, "  --temp N              temperature (default: %.1f)\n", (double)params.temp);
```
Comment (j-f1) [inline examples/common.cpp:215]: Shouldn’t the format specifier be changed to ask for a double? (Or the value changed to a float?) This happens in a bunch of places.

File: examples/main/main.cpp [hunk 4]
```diff
@@ -346,10 +347,10 @@ int main(int argc, char ** argv) {
...
         if ((int) embd_inp.size() <= n_consumed && !is_interacting) {
             // out of user input, sample next token
-            const float top_k          = params.top_k;
-            const float top_p          = params.top_p;
-            const float temp           = params.temp;
-            const float repeat_penalty = params.repeat_penalty;
+            const int top_k             = params.top_k;
+            const double top_p          = (double)params.top_p;
+            const double temp           = (double)params.temp;
+            const double repeat_penalty = (double)params.repeat_penalty;
```
Comment (j-f1) [inline examples/main/main.cpp:353]: Why not make the `params` struct have doubles?

File: examples/common.cpp (hunk 3 above)
Comment (j-f1) [reply examples/common.cpp:215]: Interesting, TIL. Thanks for explaining!

File: ggml.c [hunk 5]
```diff
@@ -1440,9 +1440,8 @@ inline static void ggml_vec_mul_f32 (const int n, float * z, const float * x, co
 inline static void ggml_vec_div_f32 (const int n, float * z, const float * x, const float * y) { for (int i = 0; i < n; ++i) z[i]  = x[i]/y[i];   }
 
 inline static void ggml_vec_dot_f32(const int n, float * restrict s, const float * restrict x, const float * restrict y) {
-    ggml_float sumf = 0.0;
-
 #ifdef GGML_SIMD
+    float sumf = 0.0f;
```
Comment (ggerganov) [inline ggml.c:1444]: Why was this changed to `float`?
We use it to accumulate the "leftovers" below

**Comment by ggerganov** [comment]
I disabled `-Wdouble-promotion` for C++ - it's too bad that `printf` does not support printing `float`.
The code becomes too cumbersome having to cast all printfs so it's not worth it

There are some code paths that haven't been updated yet, so we have to clear the remaining warnings there in `ggml.c`.

File: examples/perplexity/perplexity.cpp [hunk 6]
```diff
@@ -24,14 +24,16 @@ void perplexity(llama_context * ctx, const gpt_params & params) {
...
 
+    double nll = 0.0;
+
     fprintf(stderr, "%s : calculating perplexity over %d chunks\n", __func__, seq_count);
 
     for (int i = 0; i < seq_count; ++i) {
         int start = i * params.n_ctx;
-        int end = start + params.n_ctx - 1;
+        int end = start + params.n_ctx - 1; // TODO: this is not optimal, e.g. it makes the batch 511 instead of 512
+                                            //       it is better to always be power of 2 for better performance
```
Comment (ggerganov) [inline examples/perplexity/perplexity.cpp:36]: @glinscott 
Might want to fix this on `master`.
It's better to compute with powers of 2 for optimal performance

File: ggml.c (hunk 5 above)
Comment (ggerganov) [reply ggml.c:1441]: Good point - will change this back to `float` as you suggested then
